### ISSUE TITLE
Improve the API documentation for MultiService.addService to clarify its use in comparison to Service.setServiceParent

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5273]
|<img alt="dkdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dkdog reported|
|-|-|
|Trac ID|trac#5273|
|Type|enhancement|
|Created|2011-09-26 11:00:16Z|

Let's say

child = twisted.application.service.Service()

parent = twisted.application.service.MultiService()

child.setServiceParent(parent)

vs.

parent.addService(child)

behave differently.

The implementation of MultiService's addService does NOT make itself as the parent of child.

Meanwhile, the child, twisted.application.service.Service's setServiceParent not only adds itself to the parent via the call "addService" on the parent, it also sets its own parent attribute.

twisted.application.service.MultiService implements IServiceCollection, whose documentation says "Add a child service" for addService.

The behavior between MultiService's addService and a the Service's setServiceParent are asymmetric.

And it feels like twisted had intended to organize all services hierarchy in a tree.

If this is intended please tell me why and doc it so that ppl like me don't keep asking about it. Thanks!

Attachments:

* [addservice-doc.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4b8/4b8e55083217d6d34013b032e311ec1553122a8c/09aa8879ce36ae546072b4e2f92aceadcd70764c.patch) (2554 bytes) - added by exarkun on 2011-09-26 11:55:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5273 5273
type__enhancement enhancement
reporter__dkdog dkdog
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1317034816000000 1317034816000000
changetime__1317410577000000 1317410577000000
version__None None
owner__exarkun exarkun
cc__dkdog
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_4"></a>|@bitglue set owner to @exarkun|
|-|-|

If we take [the howto documentation](http://twistedmatrix.com/documents/current/core/howto/application.html) an an example, it only ever uses ```setServiceParent```; it never uses ```addService```. It doesn't come right out and say it, but I had inferred that ```setServiceParent``` is the right thing to do as a user of the application framework, while ```addService``` is only relevant if one is implementing an ```IServiceCollection```.

I'd say the API documentation should be more explicit in discouraging the use of ```addService``` directly. Besides the issues you mentioned, ```addService``` appears to perform no check that the child being added does not already have a parent. ```setServiceParent```, on the other hand, will disown the present parent if there is one, thus guaranteeing that the services remain a non-overlapping hierarchy.

I'm not sure if being a non-overlapping hierarchy is really a requirement, but in the absence of anything explicitly to the contrary, it's reasonable to assume it is. If I'm wrong, then that possibility (of being an overlapping hierarchy) should be more explicit.

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_5"></a>|@bitglue commented|
|-|-|

I guess I missed the part about "Only implementations of L{IService.setServiceParent} should use this method". Whoops. Looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [32713]) Improve IServiceCollection.addService API docs, clarifying when the method is useful

Author: exarkun
Reviewer: indigo
Fixes: [#5273](https://github.com/twisted/twisted/issues/5273)

`IServiceCollection.addService` is really an aid to `IService.setServiceParent`
implementations; point this out in the API documentation, so users know which
one they should be calling themselves to construct the service hierarchy.
